provider/oauth2/views.py

```diff
@@ -22,12 +20,10 @@ class OAuth2AccessTokenMixin(AccessTokenMixin):
     def get_access_token(self, request, user, scope, client):
         try:
             # Attempt to fetch an existing access token.
-            at = AccessToken.objects.get(user=user, client=client,
-                                         scope=scope, expires__gt=now())
+            at = AccessToken.objects.get(user=user, client=client, scope=scope, expires__gt=now())
         except AccessToken.DoesNotExist:
             # None found... make a new one!
             at = self.create_access_token(request, user, scope, client)
-            self.create_refresh_token(request, user, scope, at, client)
```

Comment:
I've implemented a combination of 2 and 3. Please have a look at `get_access_and_refresh_tokens()`.